src/lib/Microsoft.Health.Fhir.Ingest/Service/MultiClientEventHubProducerService.cs

```diff
@@ -0,0 +1,144 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Messaging.EventHubs;
+using Azure.Messaging.EventHubs.Producer;
+using EnsureThat;
+using Microsoft.Health.Logging.Telemetry;
+using Polly;
+
+namespace Microsoft.Health.Fhir.Ingest.Service
+{
+    public class MultiClientEventHubProducerService : IEventHubMessageService
+    {
+        private ITelemetryLogger _logger;
+        private readonly AsyncPolicy _retryPolicy;
+        private readonly IList<EventHubProducerClient> _internalClients;
+        private int _innerClientCounter;
+        private readonly int _numberOfClients;
+
+        public MultiClientEventHubProducerService(IList<EventHubProducerClient> clients, ITelemetryLogger logger)
+        {
+            _internalClients = EnsureArg.HasItems(clients, nameof(clients));
```

Comment:
The [EventhubProducerClient](https://github.com/Azure/azure-sdk-for-net/blob/86f674af36426b64af86566d24bd6806b443eb0f/sdk/eventhub/Azure.Messaging.EventHubs/src/Producer/EventHubProducerClient.cs#L334) maintains an internal pool of writers per partition. The idea of this class was to keep a collection of eventhubProducerClients and distribute requests between them, allowing us to use multiple writers per partition concurrently.